hip_data_tools/aws/s3.py

```diff
@@ -190,3 +181,239 @@ def _list_object_page(self, key_prefix, continuation_token):
+            path_in_str = str(path)
+            filename = os.path.basename(path_in_str)
+            if rename:
+                filename = f"file-{str(uuid.uuid4())}.{extension}"
+            destination_key = f"{target_key}/{filename}"
+            itr = itr + 1
+            upload_data += [(self.conn.settings, path_in_str, self.bucket, destination_key)]
+        pool_size = min(16, max(1, int(len(upload_data) / 3)))  # limit pool size between 1 and 16
+        log.debug("uploading with a multiprocessing pool of %s processes", pool_size)
+
+        Pool(pool_size).starmap(_multi_process_upload_file, upload_data)
+        log.info("Saved csv chunks at s3://%s/%s", self.bucket, target_key)
+
+    def delete_recursive_match_suffix(self, key_prefix: str, suffix: str) -> None:
+        """
+        Recursively delete all keys with given key prefix and suffix from the bucket
+        Args:
+            key_prefix (str): Key prefix under which all files will be deleted.
+            suffix (str): suffix of the subset of files in the given prefix directory to be deleted
+        Returns: None
+        """
+        if not suffix:
+            raise ValueError("suffix must not be empty")
+        s3 = self.get_resource()
+        for obj in s3.Bucket(self.bucket).objects.filter(Prefix=key_prefix):
+            if obj.key.endswith(suffix):
+                log.info("deleting s3://%s/%s", self.bucket, obj.key)
+                response = obj.delete()
+                log.info("Response: %s ", response)
+
+    def download_directory(self, source_key: str, file_suffix: str, local_directory: str) -> None:
+        """
+        Download an entire directory from s3 onto local file system
+        Args:
+            source_key (str): key prefix of the directory to be downloaded from s3
+            file_suffix (str): suffix to sunset the files to be downloaded
+            local_directory (str): local absolute path to store all the files
+        Returns: None
+        """
+        s3 = self.get_resource()
+        log.info("Downloading s3://%s/%s to %s", self.bucket, source_key, local_directory)
+        for obj in s3.Bucket(self.bucket).objects.filter(Prefix=source_key):
+            key_path = obj.key.split("/")
+            if obj.key.endswith(file_suffix):
+                filename = f"{local_directory}/{key_path[-1]}"
+                self.download_file(
+                    local_file_path=filename,
+                    key=obj.key)
+
+    def upload_json(self, key: str, json_list: List[dict]) -> None:
+        """
+        Save the json/dict data structure onto s3 as a file without using temporary local files
+        Args:
+            key: target key of the file on s3
+            json_list: a list of dictionaries that are saved as newline json in a file
+        Returns: None
+        """
+        s3 = self.get_resource()
+        s3.Object(self.bucket, key).put(
+            Body=(bytes(json.dumps(json_list, indent=2).encode('UTF-8')))
```

Comment:
what if we assign this 'utf-8' value into a constant. It helps to avoid this duplicate string value and make the code a little bit more maintainable.

Reply:
good catch